receiver/mongodbatlasreceiver/receiver.go

```diff
@@ -77,46 +81,97 @@ func (s *mongodbatlasreceiver) shutdown(context.Context) error {
-					return fmt.Errorf("error when polling process metrics from MongoDB Atlas for process %s: %w", process.ID, err)
-				}
+	for _, projectWithConfig := range projectsWithConfig {
+		if err := s.processProject(ctx, time, projectWithConfig.Project, projectWithConfig.ProjectConfig); err != nil {
+			s.log.Error("error processing project", zap.String("projectID", projectWithConfig.Project.ID), zap.Error(err))
+		}
+	}
 
-				if err := s.extractProcessDatabaseMetrics(ctx, time, org.Name, project, process, clusterName); err != nil {
-					return fmt.Errorf("error when polling process database metrics from MongoDB Atlas for process %s: %w", process.ID, err)
-				}
+	return nil
+}
 
-				if err := s.extractProcessDiskMetrics(ctx, time, org.Name, project, process, clusterName); err != nil {
-					return fmt.Errorf("error when polling process disk metrics from MongoDB Atlas for process %s: %w", process.ID, err)
-				}
-			}
+func (s *mongodbatlasreceiver) processProject(ctx context.Context, time timeconstraints, project *mongodbatlas.Project, projectCfg *ProjectConfig) error {
+	org, err := s.client.GetOrganization(ctx, project.OrgID)
+	if err != nil {
+		return fmt.Errorf("error retrieving organization: %w", err)
+	}
+
+	nodeClusterMap, err := s.getNodeClusterNameMap(ctx, project.ID)
+	if err != nil {
+		return fmt.Errorf("error collecting clusters from project %s: %w", project.ID, err)
+	}
+
+	processes, err := s.client.Processes(ctx, project.ID)
+	if err != nil {
+		return fmt.Errorf("error retrieving MongoDB Atlas processes for project %s: %w", project.ID, err)
+	}
+
+	for _, process := range processes {
+		clusterName := nodeClusterMap[process.UserAlias]
+
+		if !shouldProcessCluster(projectCfg, clusterName) {
+			// Skip processing for this cluster
+			continue
+		}
+
+		if err := s.extractProcessMetrics(ctx, time, org.Name, project, process, clusterName); err != nil {
+			return fmt.Errorf("error when polling process metrics from MongoDB Atlas for process %s: %w", process.ID, err)
+		}
+
+		if err := s.extractProcessDatabaseMetrics(ctx, time, org.Name, project, process, clusterName); err != nil {
+			return fmt.Errorf("error when polling process database metrics from MongoDB Atlas for process %s: %w", process.ID, err)
+		}
+
+		if err := s.extractProcessDiskMetrics(ctx, time, org.Name, project, process, clusterName); err != nil {
+			return fmt.Errorf("error when polling process disk metrics from MongoDB Atlas for process %s: %w", process.ID, err)
 		}
 	}
+
 	return nil
 }
 
+// shouldProcessCluster checks whether a given cluster should be processed based on the project configuration.
+func shouldProcessCluster(projectCfg *ProjectConfig, clusterName string) bool {
+	_, isIncluded := projectCfg.includesByClusterName[clusterName]
+	_, isExcluded := projectCfg.excludesByClusterName[clusterName]
+	return (isIncluded && !isExcluded) || (len(projectCfg.IncludeClusters) == 0 && !isExcluded)
```

Comment:
Feels like the logic is a little unclear here, could read a bit clearer if we reorganize the logic to represent the cases we're evaluating.
```
(len(projectCfg.IncludeClusters) == 0 || isIncluded) && !isExcluded
```

Reply:
I split the return function into 2 parts with more comments to make it more clear.